native-link-store/src/completeness_checking_store.rs

```diff
@@ -0,0 +1,362 @@
+                    }
+                    state.digests_to_check_idxs.extend(iter::repeat(i).take(rep_len));
+                    state.notify.notify_one();
+                }
+
+                // Hot path: It is very common for no output directories to be defined.
+                // So we can avoid any needless work by early returning.
+                if output_directories.is_empty() {
+                    return Ok(());
+                }
+
+                check_output_directories(cas_store, output_directories, &move |digest_infos| {
+                    let mut state = state_mux.lock();
+                    let rep_len = digest_infos.len();
+                    state.digests_to_check.extend(digest_infos);
+                    state.digests_to_check_idxs.extend(iter::repeat(i).take(rep_len));
+                    state.notify.notify_one();
+                })
+                .await?;
+
+                Result::<(), Error>::Ok(())
+            }
+            // Add a tip to the error to help with debugging and the index of the
+            // digest that failed so we know which one to unset.
+            .map_err(move |mut e| {
+                if e.code != Code::NotFound {
+                    e = e.append("Error checking existance of digest in CompletenessCheckingStore::has");
+                }
+                (e, i)
+            })
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    // This future will wait for the notify to be notified and then
+    // check the CAS store for the digest's existance.
+    // For optimization reasons we only allow one outstanding call to
+    // the underlying `has_with_results()` at a time. This is because
+    // we want to give the ability for stores to batch requests together
+    // whenever possible.
+    // The most common case is only one notify will ever happen.
+    let check_existance_fut = async {
+        let mut has_results = vec![];
+        let notify = state_mux.lock().notify.clone();
+        loop {
+            notify.notified().await;
+            let (digests, indexes) = {
+                let mut state = state_mux.lock();
+                if state.done {
+                    if state.digests_to_check.is_empty() {
+                        break;
+                    }
+                    // Edge case: It is possible for our `digest_to_check` to have had
+                    // data added, `notify_one` called, then immediately `done` set to
+                    // true. We protect ourselves by checking if we have digests if done
+                    // is set, and if we do, let ourselves know to run again, but continue
+                    // processing the data.
+                    notify.notify_one();
+                }
+                (
+                    mem::take(&mut state.digests_to_check),
+                    mem::take(&mut state.digests_to_check_idxs),
+                )
+            };
+            assert!(
+                digests.len() == indexes.len(),
```

Comment:
nit-ish: would the idiomatically Rust way to do this be something like:
```rust
assert_eq!(digests.len(), indexes.len(), 
                       "Expected sizes to match in CompletenessCheckingStore::has");
```                      
Two lines in my comment so it's not too long, but it could be one.

Reply:
Nice work overall, though. My goodness. Very complicated.